services/requester/pool.go

```diff
@@ -102,6 +154,317 @@ func (t *TxPool) Send(
 	return nil
 }
 
+// buildTransaction creates a Cadence transaction from the provided script,
+// with the given arguments and signs it with the configured COA account.
+func (t *SingleTxPool) buildTransaction(
+	ctx context.Context,
+	script []byte,
+	args ...cadence.Value,
+) (*flow.Transaction, error) {
+	defer func() {
+		t.collector.AvailableSigningKeys(t.keystore.AvailableKeys())
+	}()
+
+	var (
+		g           = errgroup.Group{}
+		err1, err2  error
+		latestBlock *flow.Block
+		account     *flow.Account
+	)
+	// execute concurrently so we can speed up all the information we need for tx
+	g.Go(func() error {
+		latestBlock, err1 = t.client.GetLatestBlock(ctx, true)
+		return err1
+	})
+	g.Go(func() error {
+		account, err2 = t.client.GetAccount(ctx, t.config.COAAddress)
+		return err2
+	})
+	if err := g.Wait(); err != nil {
+		return nil, err
+	}
+
+	flowTx := flow.NewTransaction().
+		SetScript(script).
+		SetReferenceBlockID(latestBlock.ID).
+		SetComputeLimit(flowGo.DefaultMaxTransactionGasLimit)
+
+	for _, arg := range args {
+		if err := flowTx.AddArgument(arg); err != nil {
+			return nil, fmt.Errorf("failed to add argument: %s, with %w", arg, err)
+		}
+	}
+
+	// building and signing transactions should be blocking,
+	// so we don't have keys conflict
+	t.mux.Lock()
+	defer t.mux.Unlock()
+
+	accKey, err := t.keystore.Take()
+	if err != nil {
+		return nil, err
+	}
+
+	if err := accKey.SetProposerPayerAndSign(flowTx, account); err != nil {
+		accKey.Done()
+		return nil, err
+	}
+
+	// now that the transaction is prepared, store the transaction's metadata
+	accKey.SetLockMetadata(flowTx.ID(), latestBlock.Height)
+
+	t.collector.OperatorBalance(account)
+
+	return flowTx, nil
+}
+
+type pooledEvmTx struct {
+	txPayload cadence.String
+	nonce     uint64
+}
+
+// BatchTxPool is a `TxPool` implementation that collects and groups
+// transactions based on their EOA signer, and submits them for execution
+// using a batch.
+//
+// The underlying Cadence EVM API used, is `EVM.batchRun`, instead of the
+// `EVM.run` used in `SingleTxPool`.
+//
+// The main advantage of this implementation over the `SingleTxPool`, is the
+// guarantee that transactions originated from the same EOA address, which
+// arrive in a short time interval (about the same as Flow's block productionrate),
+// will be executed in the same order their arrived.
+// This helps to reduce the nonce mismatch errors which mainly occur from the
+// re-ordering of Cadence transactions that happens from Collection nodes.
+type BatchTxPool struct {
```

Comment:
Maybe we can move it into a different file such as `services/requester/batch_tx_pool.go`